### PR TITLE
*: add TiDB Binlog Cluster doc

### PR DESCRIPTION
**Pull request by CaitinChen**
@lilin90 @WangXiangUSTC PTAL
Via 
https://github.com/pingcap/docs-cn/pull/933
https://github.com/pingcap/docs-cn/pull/961
https://github.com/pingcap/docs-cn/pull/959
https://github.com/pingcap/docs-cn/pull/973

### CODE REVIEW

File: tools/tidb-binlog-cluster.md [hunk 1]
```diff
@@ -0,0 +1,614 @@
...
+    2018/06/21 11:24:47 meta.go:117: [info] meta: &{CommitTS:400962745252184065}
+    ```
+
+    After this command is executed, a file named `{data-dir}/savepoint` is generated. This file contains `tso`, whose value is used as the value of the `initial-commit-ts` parameter needed for the initial start of Drainer.
+
+2. Backup and restore.
+
+    If the downstream is MySQL/TiDB, you need to guarantee the data integrity, and make a full backup of the data and recover the data before Drainer starts (about 10 minutes after Pump starts to run).
+
+    It is recommended to use mydumper to make a full backup of TiDB and then use Loader to export the data to the downstream. For more details, see [Backup and Restore](https://github.com/pingcap/docs/blob/master/op-guide/backup-restore.md).
```
Comment (lilin90) [inline tools/tidb-binlog-cluster.md]:
```suggestion
    It is recommended to use [mydumper](../tools/mydumper.md) to make a full backup of TiDB and then use [Loader](../tools/loader.md) to export the data to the downstream. For more details, see [Backup and Restore](../op-guide/backup-restore.md).
```

File: tools/tidb-binlog-cluster.md [hunk 2]
```diff
@@ -0,0 +1,614 @@
...
+
+3. Modify the `tidb-ansible/inventory.ini` file.
+
+    Add the deployment machine IPs for `drainer_servers`. Set `initial_commit_ts` to the value you have obtained, which is used just for the initial start of Drainer.
```
Comment (lilin90) [inline tools/tidb-binlog-cluster.md]:
```suggestion
    Add the deployment machine IPs for `drainer_servers`. Set `initial_commit_ts` to the value you have obtained, which is only used for the initial start of Drainer.
```

File: tools/tidb-binlog-cluster.md [hunk 3]
```diff
@@ -0,0 +1,614 @@
...
+
+    - Assume that the downstream is MySQL with its alias `drainer_mysql`:
```
Comment (lilin90) [inline tools/tidb-binlog-cluster.md]:
```suggestion
    - Assume that the downstream is MySQL with the alias `drainer_mysql`:
```

File: tools/tidb-binlog-cluster.md [hunk 4]
```diff
@@ -0,0 +1,614 @@
...
+
+        ```
+        [drainer_servers]
+        drainer_mysql ansible_host=172.16.10.71 initial_commit_ts="402899541671542785"
+        ```
+
+    - Assume that the downstream is `pb` with its alias `drainer_pb`:
```
Comment (lilin90) [inline tools/tidb-binlog-cluster.md]:
```suggestion
    - Assume that the downstream is `pb` with the alias `drainer_pb`:
```

File: tools/tidb-binlog-cluster.md [hunk 5]
```diff
@@ -0,0 +1,614 @@
...
+
+        ```
+        [drainer_servers]
+        drainer_pb ansible_host=172.16.10.71 initial_commit_ts="402899541671542785"
+        ```
+
+4. Modify the configuration file.
+
+    1. Assume that the downstream is MySQL:
```
Comment (lilin90) [inline tools/tidb-binlog-cluster.md]: A bulleted list is more appropriate for this section.

File: tools/tidb-binlog-cluster.md [hunk 6]
```diff
@@ -0,0 +1,614 @@
...
+        # dir = "data.drainer"
+        ```
+
+5. Deploy Drainer.
+
+    ```
+    $ ansible-playbook deploy_drainer.yml
+    ```
+
+6. Starty Drainer.
```
Comment (lilin90) [inline tools/tidb-binlog-cluster.md]:
```suggestion
6. Start Drainer.
```

File: tools/tidb-binlog-cluster.md [hunk 7]
```diff
@@ -0,0 +1,614 @@
...
+wget https://download.pingcap.org/tidb-binlog-cluster-latest-linux-amd64.tar.gz
+wget https://download.pingcap.org/tidb-binlog-cluster-latest-linux-amd64.sha256
+
+# Check the file integrity. It should return OK.
+sha256sum -c tidb-binlog-cluster-latest-linux-amd64.sha256
+```
+
+#### The usage example
+
+Assume that there are three PDs, one TiDB, two machines used for deploying Pump, and one machine used for deploying Drainer. The information for each node is as follows:
```
Comment (lilin90) [inline tools/tidb-binlog-cluster.md]:
```suggestion
Assuming that you have three PD nodes, one TiDB node, two Pump nodes, and one Drainer node, the information of each node is as follows:
```

File: tools/tidb-binlog-cluster.md [hunk 8]
```diff
@@ -0,0 +1,614 @@
...
+TiDB="192.168.0.10"
+PD1="192.168.0.16"
+PD2="192.168.0.15"
+PD3="192.168.0.14"
+Pump="192.168.0.11"
+Pump="192.168.0.12"
+Drainer="192.168.0.13"
+```
+
+For this example, the following shows how to use Pump/Drainer.
```
Comment (lilin90) [inline tools/tidb-binlog-cluster.md]:
```suggestion
The following part shows how to use Pump and Drainer based on the nodes above.
```

File: tools/tidb-binlog-cluster.md [hunk 9]
```diff
@@ -0,0 +1,614 @@
...
+        -L string
+            the output information level of logs: debug, info, warn, error, fatal ("info" by default)
+        -V
+            the print version information
+        -addr string
+            the RPC address through which Pump provides the service (-addr="192.168.0.11:8250")
+        -advertise-addr string
+            the RPC address through which Pump provides the external service (-advertise-addr="192.168.0.11:8250")
+        -config string
+            the configuration file directory. If you have specified the configuration file, Pump reads the configuration file first. If the corresponding configuration exits in the command line parameters, Pump uses the configuration of the command line parameters to cover that of the configuration file.
```
Comment (lilin90) [inline tools/tidb-binlog-cluster.md]:
```suggestion
            the path of the configuration file. If you specify the configuration file, Pump reads the configuration in the configuration file first. If the corresponding configuration also exits in the command line parameters, Pump uses the configuration of the command line parameters to cover that of the configuration file.
```

File: tools/tidb-binlog-cluster.md [hunk 10]
```diff
@@ -0,0 +1,614 @@
...
+        -data-dir string
+            the directory where the Pump data is stored
```
Comment (lilin90) [inline tools/tidb-binlog-cluster.md]:
```suggestion
            the path where the Pump data is stored
```

File: tools/tidb-binlog-cluster.md [hunk 11]
```diff
@@ -0,0 +1,614 @@
...
+        -enable-tolerant
+            After `tolerant` is enabled, Pump (enabled by default) does not report an error if it fails to write in the binlog.
```
Comment (lilin90) [inline tools/tidb-binlog-cluster.md]:
```suggestion
            After `tolerant` is enabled, Pump (enabled by default) does not report an error if the binlog fails to write into Pump.
```

File: tools/tidb-binlog-cluster.md [hunk 12]
```diff
@@ -0,0 +1,614 @@
...
+        -gc int
+            the days of data which Pump saves (7 by default)
```
Comment (lilin90) [inline tools/tidb-binlog-cluster.md]:
```suggestion
            Pump retains the data within how many days (7 by default)
```

File: tools/tidb-binlog-cluster.md [hunk 13]
```diff
@@ -0,0 +1,614 @@
...
+        -heartbeat-interval int
+            the interval of the heartbeats Pump sends to PD (in seconds)
+        -log-file string
+            log file directory
```
Comment (lilin90) [inline tools/tidb-binlog-cluster.md]:
```suggestion
            the file path of logs
```

File: tools/tidb-binlog-cluster.md [hunk 14]
```diff
@@ -0,0 +1,614 @@
...
+        
+        > **Note:** If the downstream is MySQL/TiDB, to guarantee the data integrity, you need to obtain the `initial-commit-ts` value and make a full backup of the data and restore the data before the initial start of Drainer. For details, see [Deploy Drainer](#deploy-drainer).
+    
+        When Pump is started for the first time, use the `initial-commit-ts` parameter.
+
+        ```bash
+        ./bin/drainer -config drainer.toml -initial-commit-ts {initial-commit-ts}
+        ```
+
+        If the command line parameter and the configuration file parameter are the same, use the parameter value in the command line.
```
Comment (lilin90) [inline tools/tidb-binlog-cluster.md]:
```suggestion
        If the command line parameter and the configuration file parameter are the same, the parameter value in the command line is used.
```

File: tools/tidb-binlog-cluster.md [hunk 15]
```diff
@@ -0,0 +1,614 @@
...
+For how to pause, close, and check and modify the state of Drainer, see the [binlogctl guide](#binlogctl-guide) as follows.
+
+### binlogctl guide
+
+[binlogctl](https://github.com/pingcap/tidb-tools/tree/develop/tidb-binlog/binlogctl) is an operations tool for TiDB-Binlog with the following features:
+
+* Obtains the current `ts`
+* Checks the Pump/Drainer state
+* Modifies the Pump/Drainer state
+* Pause/Close Pump/Drainer
```
Comment (lilin90) [inline tools/tidb-binlog-cluster.md]:
```suggestion
* Pausing or closing Pump/Drainer
```

File: tools/tidb-binlog-cluster.md [hunk 16]
```diff
@@ -0,0 +1,614 @@
...
+  * The binlog data which Pump saves is processed by GC only when it is consumed by all the Drainers that are not `offline`.
+  * Close Drainer only when it will not be used any more.
+
+For how to pause, close, and check and modify the state of Drainer, see the [binlogctl guide](#binlogctl-guide) as follows.
+
+### binlogctl guide
+
+[binlogctl](https://github.com/pingcap/tidb-tools/tree/develop/tidb-binlog/binlogctl) is an operations tool for TiDB-Binlog with the following features:
+
+* Obtains the current `ts`
```
Comment (lilin90) [inline tools/tidb-binlog-cluster.md]:
```suggestion
* Obtaining the current `ts`
```

File: tools/tidb-binlog-cluster.md [hunk 17]
```diff
@@ -0,0 +1,614 @@
...
+* Checks the Pump/Drainer state
```
Comment (lilin90) [inline tools/tidb-binlog-cluster.md]:
```suggestion
* Checking the Pump/Drainer state
```

File: tools/tidb-binlog-cluster.md [hunk 18]
```diff
@@ -0,0 +1,614 @@
...
+* Modifies the Pump/Drainer state
```
Comment (lilin90) [inline tools/tidb-binlog-cluster.md]:
```suggestion
* Modifying the Pump/Drainer state
```

File: tools/tidb-binlog-cluster.md [hunk 19]
```diff
@@ -0,0 +1,614 @@
...
+* Pause/Close Pump/Drainer
+
+#### Usage scenarios of binlogctl
+
+* The first time you run Drainer and need to obtain the current `ts`
```
Comment (lilin90) [inline tools/tidb-binlog-cluster.md]:
```suggestion
* It is the first time you run Drainer and you need to obtain the current `ts`.
```

File: tools/tidb-binlog-cluster.md [hunk 20]
```diff
@@ -0,0 +1,614 @@
...
+* When Pump/Drainer exits abnormally, its state is not updated and the service is affected
```
Comment (lilin90) [inline tools/tidb-binlog-cluster.md]:
```suggestion
* When Pump/Drainer exits abnormally, its state is not updated and the service is affected. You can use this tool to modify the state.
```

File: tools/tidb-binlog-cluster.md [hunk 21]
```diff
@@ -0,0 +1,614 @@
...
+* When a fault occurs in synchronization or you need to check the runtime, and you need to check the Pump/Drainer state
```
Comment (lilin90) [inline tools/tidb-binlog-cluster.md]:
```suggestion
* An error occurs during synchronization and you need to check the running status and the Pump/Drainer state.
```

File: tools/tidb-binlog-cluster.md [hunk 22]
```diff
@@ -0,0 +1,614 @@
...
+* When you need to maintain the cluster and should pause/close Pump/Drainer
```
Comment (lilin90) [inline tools/tidb-binlog-cluster.md]:
```suggestion
* While maintaining the cluster, you need to pause or close Pump/Drainer.
```

File: tools/tidb-binlog-cluster.md [hunk 23]
```diff
@@ -0,0 +1,614 @@
...
+
+```bash
+wget https://download.pingcap.org/binlogctl-new-linux-amd64.tar.gz
+wget https://download.pingcap.org/binlogctl-new-linux-amd64.sha256
+
+# Check the file integrity. It should return OK.
+sha256sum -c tidb-binlog-new-linux-amd64.sha256
+```
+
+#### How to use binlogctl
```
Comment (lilin90) [inline tools/tidb-binlog-cluster.md]:
```suggestion
#### binlogctl usage description
```

File: tools/tidb-binlog-cluster.md [hunk 24]
```diff
@@ -0,0 +1,614 @@
...
+```
+
+#### How to use binlogctl 
+
+Command line parameters:
+
+```
+Usage of binlogctl:
+-V	
+Outputs the binlogctl version information
```
Comment (lilin90) [inline tools/tidb-binlog-cluster.md]: Keep the indentation consistent for the description lines in this code block.

File: tools/tidb-binlog-cluster.md [hunk 25]
```diff
@@ -0,0 +1,614 @@
...
+-cmd string
+    the command mode, including "generate_meta", "pumps", "drainers", "update-pump" ,"update-drainer", "pause-pump", "pause-drainer", "offline-pump", and "offline-drainer"
+-data-dir string
+    the directory where the checkpoint file of Drainer is stored ("binlog_position" by default)
```
Comment (lilin90) [inline tools/tidb-binlog-cluster.md]:
```suggestion
    the file path where the checkpoint file of Drainer is stored ("binlog_position" by default)
```

File: tools/tidb-binlog-cluster.md [hunk 26]
```diff
@@ -0,0 +1,614 @@
...
+-node-id string
+    ID of Pump/Drainer
+-pd-urls string
+    the address of PD. If multiple addresses exist, use "," to separate each ("http://127.0.0.1:2379" by default)
+-ssl-ca string
+    the directory of SSL CAs
```
Comment (lilin90) [inline tools/tidb-binlog-cluster.md]:
```suggestion
    the file path of SSL CAs
```

File: tools/tidb-binlog-cluster.md [hunk 27]
```diff
@@ -0,0 +1,614 @@
...
+-ssl-cert string
+        the directory of X509 certificate file of the PEM format
```
Comment (lilin90) [inline tools/tidb-binlog-cluster.md]:
```suggestion
        the file path of the X509 certificate file in the PEM format
```

File: tools/tidb-binlog-cluster.md [hunk 28]
```diff
@@ -0,0 +1,614 @@
...
+-ssl-key string
+        the directory of X509 key file of the PEM format
```
Comment (lilin90) [inline tools/tidb-binlog-cluster.md]: Ditto

File: tools/tidb-binlog-cluster.md [hunk 29]
```diff
@@ -0,0 +1,614 @@
...
+-time-zone string
+    If a time zone is set, the corresponding time of the obtained `tso` is printed in the "generate_meta" mode. For example, "Asia/Shanghai" is the CST time zone and "Local" is the local time zone
+```
+
+Command example:
+
+- Check the state of all the Pump/Drainers:
```
Comment (lilin90) [inline tools/tidb-binlog-cluster.md]:
```suggestion
- Check the state of all the Pumps or Drainers:
```

File: tools/tidb-binlog-cluster.md [hunk 30]
```diff
@@ -0,0 +1,614 @@
...
+
+    ```
+    The Pump/Drainer states include `online`, `pausing`, `paused`, `closing` and `offline`. 
+
+    bin/binlogctl -pd-urls=http://127.0.0.1:2379 -cmd update-pump/update-drainer -node-id ip-127-0-0-1:8250/{nodeID} -state {state}
+    ```
+
+    This command modifies the Pump/Drainer state saved in PD.
+ 
+- Pause/Close Pump/Drainer:
```
Comment (lilin90) [inline tools/tidb-binlog-cluster.md]:
```suggestion
- Pause or close Pump/Drainer:
```

File: tools/tidb-binlog-cluster.md [hunk 31]
```diff
@@ -0,0 +1,614 @@
...
+
+    ```
+    bin/binlogctl -pd-urls=http://127.0.0.1:2379 -cmd pause-pump/pause-drainer/offline-pump/offline-drainer -node-id ip-127-0-0-1:8250/{nodeID}
+    ```
+
+    binlogctl sends the HTTP request to Pump/Drainer, and Pump/Drainer exits from the process after receiving the command and sets its state to `paused`/`offline`.
+
+- Generate the meta file which is needed for the Drainer startup:
```
Comment (lilin90) [inline tools/tidb-binlog-cluster.md]:
```suggestion
- Generate the meta file that Drainer needs to start:
```

**Comment by CaitinChen** [comment]
@lilin90 Thanks. PTAL again.

File: tools/tidb-binlog-cluster.md [hunk 32]
```diff
@@ -0,0 +1,620 @@
+---
+title: TiDB-Binlog Cluster User Guide
+summary: This document introduces TiDB-Binlog of the cluster version, including its architecture, deployment, operations and monitoring.
```
Comment (lilin90) [inline tools/tidb-binlog-cluster.md]: Please refer to other documents for the summary sentence pattern. Only several specific sentence patterns are used.

File: tools/tidb-binlog-cluster.md [hunk 33]
```diff
@@ -0,0 +1,620 @@
...
+Pump/Drainer state description:
+
+* `online`: running normally
+* `pausing`: in the pausing process. It turns into this state after you use `kill` or press Ctrl + C to exit from the process.
+* `paused`: has been stopped. While Pump is in this state, it rejects the request of writing binlog into it and does not provide the binlog for Drainer any more. When Drainer is in this state, it does not synchronize data to the downstream. After Pump and Drainer exit normally from all the threads, they switch the state to `paused` and then exits from the process.
+* `closing`: in the offline process. `binlogctl` is used to get Pump/Drainer offline and Pump/Drainer is in this state before the process exits. In this state, Pump does not accept new requests of writing binlog into it and waits for all the binlog data to be used up by Drainer.
+* `offline`: becomes offline. After Pump sents all the binlog data that it saves to Drainer, its state is switched to `offline`. Drainer's state can be switched to `offline` after all the threads have exited.
+
+> **Notes:**
+
```
Comment (lilin90) [inline tools/tidb-binlog-cluster.md]: This blank line also needs ">".

File: tools/tidb-binlog-cluster.md [hunk 34]
```diff
@@ -0,0 +1,620 @@
...
+    2018/06/21 11:24:47 meta.go:117: [info] meta: &{CommitTS:400962745252184065}
+    ```
+
+    This command generates a `{data-dir}/savepoint` file. This file stores the `tso` information which is needed for the initial start of Drainer.
+
+## TiDB-Binlog monitoring
+
+After you have deployed TiDB-Binlog using Ansible successfully, you can go to the Grafana Web (default address: <http://grafana_ip:3000>, default account: admin, password: admin) to check the state of Pump and Drainer.
+
+For TiDB-Binlog monitoring metrics, see [TiDB-Binlog Monitoring Metrics](../tools/tidb-binlog-monitor.md)
```
Comment (lilin90) [inline tools/tidb-binlog-cluster.md]: Add a period in the end.

File: tools/tidb-binlog-cluster.md [hunk 35]
```diff
@@ -0,0 +1,613 @@
...
+
+The TiDB-Binlog architecture is as follows:
+
+![TiDB-Binlog architecture](../media/tidb_binlog_cluster_architecture.png)
+
+The TiDB-Binlog cluster is composed of Pump and Drainer.
+
+### Pump
+
+Pump is used to record the binlogs generated in TiDB, reorder the binlogs based on the commit time of the transaction, and send binlogs to Drainer for consumption.
```
[the remark this answers is not in the document]
Comment (WangXiangUSTC) [reply tools/tidb-binlog-cluster.md]: I think sort is better

File: tools/tidb-binlog-cluster.md [hunk 36]
```diff
@@ -0,0 +1,613 @@
...
+* Binlog is started by adding the `enable-binlog` startup parameter to TiDB.
+* Drainer does not support the `rename` DDL operation on the table of `ignore schemas` (the schemas in the filter list).
+* If you want to start Drainer in the existing TiDB cluster, generally, you need to make a full backup of the cluster data, obtain `savepoint`, export the data to the target database, and start Drainer to synchronize the incremental data from `savepoint`.
+* Drainer supports synchronizing binlogs to MySQL, TiDB, Kafka or the local files. If you need to synchronize binlogs to other destinations, you can set Drainer to synchronize the binlog to Kafka and read the data in Kafka for customization processing. See [Binlog Slave Client User Guide](../tools/binlog-slave-client.md).
+* Pump/Drainer has two states: `paused` and `offline`. If you press Ctrl + C or kill the process, both Pump and Drainer become `paused`. There is no need for paused Pump to send all the binlog data to Drainer. If you need to exit from Pump for long (or do not use Pump any more), you need to use binlogctl to turn Pump offline. The same goes for Drainer.
+
+## TiDB-Binlog deployment
+
```
[the remark this answers is not in the document]
Comment (lilin90) [reply tools/tidb-binlog-cluster.md]: @WangXiangUSTC Got it! @CaitinChen Please add the related description here.

File: tools/tidb-binlog-cluster.md (hunk 35 above)
[the remark this answers is not in the document]
Comment (lilin90) [reply tools/tidb-binlog-cluster.md]: @CaitinChen Please update the several "reorder"s to "sort". Thanks!

**Comment by CaitinChen** [comment]
@lilin90 PTAL again. Thanks.

**Comment by lilin90** [comment]
@CaitinChen Please check and update two other comments about "reorder" and adding the description of two methods.

**Comment by CaitinChen** [comment]
@lilin90 I have updated the doc. PTAL. Thanks.

File: tools/tidb-binlog-cluster.md [hunk 37]
```diff
@@ -0,0 +1,622 @@
...
+The TiDB-Binlog cluster is composed of Pump and Drainer.
+
+### Pump
+
+Pump is used to record the binlogs generated in TiDB, sort the binlogs based on the commit time of the transaction, and send binlogs to Drainer for consumption.
+
+### Drainer
+
+Drainer collects and merges binlogs from each Pump, converts the binlog to SQL or data of a specific format, and synchronizes the data to the downstream.
```
Comment (lilin90) [inline tools/tidb-binlog-cluster.md]:
```suggestion
Drainer collects and combines binlogs from each Pump, converts the binlog to SQL or data of a specific format, and synchronizes the data to the downstream.
```

File: tools/tidb-binlog-cluster.md [hunk 38]
```diff
@@ -0,0 +1,622 @@
...
+
+## Main features
+
+* Multiple Pumps form a cluster which can scale out horizontally.
+* TiDB uses the built-in Pump Client to send the binlog to each Pump.
+* Pump stores binlogs and sends the binlogs to Drainer in order.
+* Drainer reads binlogs of each Pump, merges and reorders the binlogs, and sends the binlogs to the downstream.
```
Comment (lilin90) [inline tools/tidb-binlog-cluster.md]:
```suggestion
* Drainer reads binlogs of each Pump, combines and sorts the binlogs, and sends the binlogs to the downstream.
```

**Comment by CaitinChen** [comment]
@lilin90  Thanks ~  PTAL again.

**Comment by CaitinChen** [comment]
@WangXiangUSTC PTAL

**Comment by GregoryIan** [comment]
any updates?

**Comment by CaitinChen** [comment]
@GregoryIan Merged. Any comments?

**Comment by GregoryIan** [comment]
no, thanks, we need it in release note

**Pull request merged**